### DISCUSSION TITLE
is it possible to keep the chunk safely after publish it?

### DISCUSSION THREAD

**Discussion by wxtry** [eclipse-iceoryx/iceoryx#981] [Q&A] [answered]
I'm trying to maintanance several pieceses of shared memory between publisher and subscribers.
I want the publisher can allocate a piece of memory then publish and manupilate it repeatedly any time it wants, any subscriber can get the same memory on each publish call without any problem.
But in version 1.0.1, the chunk being sent has to be released from the publisher side in `ChunkSender<ChunkSenderDataType>::getChunkReadyForSend` :
```
    if (getMembers()->m_chunksInUse.remove(chunkHeader, chunk))
    {
        chunk.getChunkHeader()->setSequenceNumber(getMembers()->m_sequenceNumber++);
        return true;
    }
```
So re-publish and edit to this chunk safely is no more possbile.
I'm curious if it is possible to send the chunk and keep reference on publisher side for further manupilation ?
If the anwer is negative, in what condition this could be harmful?

**Comment by elfenpiff** [comment, marked as answer]
@wxtry The answer is currently that we do not support keeping the sample after publishing.
The problems can result in a data race when some subscriber would like to read the sample while you are modifying it on the publisher side. Here we would require a mechanism to either provide exclusive read or write access to either the publisher or the subscriber. 

Normally, one would implement such a thing with a mutex but just imagine what happens when a subscriber dies while holding the mutex. Then the subscriber can deadlock the publisher as well which violates our freedom of interference principle which is required in a safety critical environment. So the solution will be a little bit more complex.

But you are lucky, we already have an issue (https://github.com/eclipse-iceoryx/iceoryx/issues/927) for this since you are not the only one with that use case.
Can you please take a look at this issue and if this does not solve your use case you can draft them in there and we try to consider it in our implementation.

**Reply by wxtry** [reply]
Thank you for the reply, it's nice to see this is already under discussion. I will keep track of the issue you provided.
